### PR TITLE
Necromancer Healing in Combat-Trainer

### PR DESCRIPTION
**Pull request by jandersson**
This is a draft of a working healing implementation. Would appreciate any feedback on how to improve. I copied a lot of the logic from HealMe.lic. I'm going to fill out the rest of the severity hash, update the vitality list, deal with nerve damage, and try to make the spellcasting logic more robust. I'm concerned the check_wounds method is in the wrong process (should I try to move it to safety process?).

### CODE REVIEW

File: combat-trainer.lic [hunk 1]
```diff
@@ -246,10 +249,29 @@ class LootProcess
...
+          bput("prepare cf", @settings.prep_messages)
+        end
+      end
+      if @ritual_type == ('harvest' || 'consume')
         check_thanatology(DRRoom.dead_npcs.first, 'preserve', game_state)
       end
       check_thanatology(DRRoom.dead_npcs.first, @ritual_type, game_state)
+      if @heal && @ritual_type == 'consume'
+        pause 0.5 until Flags['necro-spellcast']
+        fput("cast #{wounds.max_by{|k,v| v}[0]}")
```
Comment (rpherbig) [inline combat-trainer.lic]: Is this the same as `wounds.values.max.first`?

File: combat-trainer.lic [hunk 2]
```diff
@@ -246,10 +249,29 @@ class LootProcess
...
+          bput('release', '')
+          bput("prepare cf", @settings.prep_messages)
+        end
+      end
+      if @ritual_type == ('harvest' || 'consume')
         check_thanatology(DRRoom.dead_npcs.first, 'preserve', game_state)
       end
       check_thanatology(DRRoom.dead_npcs.first, @ritual_type, game_state)
+      if @heal && @ritual_type == 'consume'
+        pause 0.5 until Flags['necro-spellcast']
```
Comment (rpherbig) [inline combat-trainer.lic]: Lich has `waitcastrt?` if you want to wait for a spell to be fully prepped (i.e. the blue bar in StormFront).

File: combat-trainer.lic [hunk 3]
```diff
@@ -286,6 +308,68 @@ class LootProcess
     dispose_trash(checkright) if snap.last != checkright
   end
 
+  def check_wounds
```
Comment (rpherbig) [inline combat-trainer.lic]: If most of this logic is shared with `healme`, I'd like to see you either leverage that script, or move this code to a shared location (a new `common-*` file?). What do you think?

File: combat-trainer.lic (hunk 1 above)
Comment (jandersson) [inline combat-trainer.lic]: it returns the key and the value, so 0 gets the key. I pulled it from this: http://stackoverflow.com/questions/6040494/how-to-find-the-key-of-the-largest-value-hash

File: combat-trainer.lic [hunk 4]
```diff
@@ -246,10 +249,29 @@ class LootProcess
     game_state.mob_died = true
     arrange_mob(DRRoom.dead_npcs.first, game_state)
     if @necro
-      if @ritual_type == 'harvest'
+      if @heal
+        wounds = check_wounds
+        echo "Wounds: #{wounds}" if $debug_mode_ct
+        unless wounds.empty?
+          Flags.add('necro-spellcast', '^You feel fully prepared to cast your spell\.')
+          prev_ritual = @ritual_type
```
Comment (rpherbig) [inline combat-trainer.lic]: I'm not a fan of this logic, but I'd have to grok the code better to articulate why. Can the consume code be commonized and called from here instead of changing @ritual_type?

File: combat-trainer.lic (hunk 1 above)
Comment (rpherbig) [inline combat-trainer.lic]: Ah, thank you. What about `wounds.key(wounds.values.max)`?

Although if you're sorting by the value to find a key, it makes me think that you should consider swapping your keys and values in the hash.

File: combat-trainer.lic (hunk 3 above)
Comment (jandersson) [inline combat-trainer.lic]: Yes, im uncomfortable with this being in the looting process

File: combat-trainer.lic (hunk 4 above)
Comment (jandersson) [inline combat-trainer.lic]: Absolutely. From the new common-* file? (common-healing?)

File: combat-trainer.lic (hunk 1 above)
Comment (jandersson) [inline combat-trainer.lic]: Yea that makes sense, ill swap them

File: combat-trainer.lic (hunk 4 above)
Comment (rcuhljr) [inline combat-trainer.lic]: If we make a new common-file remember to add it to `bootstrap.lic`.

**Comment by jandersson** [comment]
Running into an issue where I interrupt `SpellProcess` by releasing its currently prepped spell. Then I prep CF, cast, and heal. `SpellProcess` tries to cast its spell which I've released and times out for 30 seconds. 
```
[combat-trainer: action count: 5 vs 10]
[combat-trainer: skill exp: 18 vs 19]
[combat-trainer]>cast
You don't have a spell prepared!
>
* Trying weakly, a young ogre bashes a spiked mace at you.  You turn aside little of the mace with a polished steel parry stick.  
[You're bruised, nimbly balanced with opponent in good position.]
>
The young ogre moves into a position to dodge.
>
* Weakly, a young ogre bashes a spiked mace at you.  You turn aside little of the mace with a polished steel parry stick.  
[You're bruised, solidly balanced with opponent in good position.]
>
You feel fully rested.
>
The young ogre moves into a position to parry.
>
A young ogre stomps around snorting.
>
[combat-trainer: *** No match was found after 15 seconds, dumping info]
[combat-trainer: messages seen length: 12]
[combat-trainer: message: You don't have a spell prepared!]
[combat-trainer: message: * Trying weakly, a young ogre bashes a spiked mace at you.  You turn aside little of the mace with a polished steel parry stick.  ]
[combat-trainer: message: [You're bruised, nimbly balanced with opponent in good position.]]
[combat-trainer: message: The young ogre moves into a position to dodge.]
[combat-trainer: message: Manifest Force  (13 roisaen)]
[combat-trainer: message: Obfuscation  (4 roisaen)]
[combat-trainer: message: * Weakly, a young ogre bashes a spiked mace at you.  You turn aside little of the mace with a polished steel parry stick.  ]
[combat-trainer: message: [You're bruised, solidly balanced with opponent in good position.]]
[combat-trainer: message:  * Monk Arimane Bloodmoon joins the adventure.]
[combat-trainer: message: You feel fully rested.]
[combat-trainer: message: The young ogre moves into a position to parry.]
[combat-trainer: message: A young ogre stomps around snorting.]
[combat-trainer: checked against [/^You gesture/i, /You reach with both/i, /You listen intently to the air/i, /You close your eyes/i, /You close yours eyes/i, /Your target pattern dissipates/i, /You make a holy gesture/i, /You raise your left hand skyward/i, /Flickering shadows tear away/i, /You sense a Malediction/i, /You release an accompaniment/i, /You roll your hands in an elliptical pattern/i, /You cup your hand/i, /You raise your right hand skyward/i, /You flicker your malevolent desires towards/i, /You clench your hands into fists and grit your teeth/i, /You make a sweeping gesture with your arm/i, /You wave your hand/i, /Tendrils of flame dart along your hand/i, /The flames dancing along your fingertips/i, /^You can't cast that at yourself/i, /You breathe deeply, centering your energies/i]]
```

**Comment by jandersson** [comment]
using `game_state.casting = false` and `game_state.cast_timer = nil` seemed to do the trick

File: combat-trainer.lic [hunk 5]
```diff
@@ -284,6 +276,21 @@ class LootProcess
...
+  def consume(mob_noun, game_state, wounds)
+    check_thanatology(DRRoom.dead_npcs.first, 'consume', game_state)
+    # Flags.add('necro-spellcast', '^You feel fully prepared to cast your spell\.')
+    if game_state.casting
+      bput('release spell', '')
+      game_state.casting = false
+    end
+    bput("prepare cf", @settings.prep_messages)
+    game_state.casting = true
+    waitcastrt?
```
Comment (rpherbig) [inline combat-trainer.lic]: How long does this wait typically take? Are you ok not taking any other actions in the meantime?

Since you've released the currently casting spell and inserted consume flesh (you should also insert a custom cast message), an alternative would be to exit the Process here and let SpellProcess take over the spellcasting. That would let other actions be taken while you wait for the spell to prep.

Comment (jandersson) [inline combat-trainer.lic]: I'd guess about 15 seconds. It seems like a good idea to have SpellProcess handle the casting.

Comment (rpherbig) [inline combat-trainer.lic]: In that case, you may want to add a check here to NOT release if the current spell is consume flesh.

File: profiles/Jonas-setup.yaml [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+
+skinning:
+  skin: true
+  arrange_count: 5
+  tie_bundle: true
+
+# Necromancer Related
+necromancer_healing:
+  CF:
+  - 15
```
Comment (rpherbig) [inline profiles/Jonas-setup.yaml]: I would highly encourage you to use the same format here as for other spells.

File: profiles/Jonas-setup.yaml [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+crossing_training:
+- Appraisal
+- Locksmithing
+- First Aid
+- Augmentation
+- Utility
+- Warding
+exp_timers:
+  Stealth:
+    cooldown: 300
```
Comment (rpherbig) [inline profiles/Jonas-setup.yaml]: It's unrelated to the PR, but you can simplify this to `Stealth: 300`

File: profiles/Jonas-setup.yaml (hunk 6 above)
Comment (rcuhljr) [inline profiles/Jonas-setup.yaml]: It's how empath healing is setup, which means we should update empath healing as well. And maybe unify training spells and combat spells.

Comment (jandersson) [inline profiles/Jonas-setup.yaml]: Yea I basically just looked at how healing was done with VH in `check_health` in SpellProcess

Comment (rpherbig) [inline profiles/Jonas-setup.yaml]: Alright, that's outside the scope of this PR then. But please do make an issue to update this and empath healing and we can get to it later.

File: profiles/Jonas-setup.yaml (hunk 7 above)
Comment (jandersson) [inline profiles/Jonas-setup.yaml]: Always good to know these little things!

File: profiles/ANecromancer-setup.yaml [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+training_manager_hunting_priority: true
+training_manager_priority_skills:
+- Parry Ability
+hunting_info:
+- :zone: young_ogres
+  stop_on:
+  - Parry Ability
+  args:
+  - d0
+  - thanatology
```
Comment (rpherbig) [inline profiles/ANecromancer-setup.yaml]: Do you have a file named `ANecromancer-thanatology.yaml`?

Comment (jandersson) [inline profiles/ANecromancer-setup.yaml]: Yes, I'll commit it

**Comment by jandersson** [comment]
Well I made a mess again by committing the wrong file. Will fix

**Comment by jandersson** [comment]
I took out the line which causes LootProcesses to wait until the room is free of dead NPC's before continuing. Now we get about 4-5 failed loot attempts while waiting for the dead NPC to decay.

```

[combat-trainer]>loot
You feel fully rested.
>
A giant black leucro skulks menacingly into view.
>
You search the black leucro.
You find nothing of interest.
>
[combat-trainer]>loot
I could not find what you were referring to.
>
[combat-trainer]>loot
I could not find what you were referring to.
>
[combat-trainer]>loot
I could not find what you were referring to.
>
[combat-trainer]>loot
You feel fully prepared to cast your spell.
>
```

**Comment by rpherbig** [comment]
You might need to add a timer so that it only attempts to loot every... 5 seconds?

**Comment by jandersson** [comment]
Alright, I will re-add the line in the meantime.

File: combat-trainer.lic [hunk 9]
```diff
@@ -607,20 +610,21 @@ class SpellProcess
   def check_charging?(game_state)
     return false unless game_state.charge_cambrinth?
 
-    find_cambrinth3(@cambrinth, @held_cambrinth)
```
Comment (jandersson) [inline combat-trainer.lic]: I don't recall doing anything with some of these lines... is this a product of merging?

Comment (rpherbig) [inline combat-trainer.lic]: Yes. I moved `find_cambrinth` and `stow_cambrinth` outside of the if. Make sure you don't revert the change in this PR.

File: combat-trainer.lic [hunk 10]
```diff
@@ -738,14 +742,13 @@ class SpellProcess
                    .select { |spell| spell['expire'] ? Flags["ct-#{spell['abbrev']}"] : true }
                    .select { |spell| game_state.dancing? ? spell['harmless'] : true }
                    .select { |spell| spell['recast_every'] ? check_spell_timer(spell) : true }
-                   .select { |spell| @cast_only_to_train ? DRSkill.getxp(spell['skill']) <= 32 : true }
     data = if @offensive_spell_cycle.empty?
              ready_spells.min_by { |spell| DRSkill.getxp(spell['skill']) }
            else
-             name = @offensive_spell_cycle.find { |spell_name| ready_spells.find { |spell| spell['name'] == spell_name } }
+             name = @offensive_spell_cycle.find { |name| ready_spells.find { |spell| spell['name'] == name } }
```
Comment (rpherbig) [inline combat-trainer.lic]: This is another merge issue.

File: combat-trainer.lic [hunk 11]
```diff
@@ -738,14 +742,13 @@ class SpellProcess
...
              @offensive_spell_cycle.rotate!
              ready_spells.find { |spell| spell['name'] == name }
-           end
+          end
```
Comment (rpherbig) [inline combat-trainer.lic]: This is another merge issue.

**Comment by rpherbig** [comment]
Do you think the loot changes you're making will fix #591?

File: combat-trainer.lic (hunk 9 above)
Comment (jandersson) [inline combat-trainer.lic]: Hmm OK, I will try merging again

**Comment by jandersson** [comment]
Yea they should, the pausing until the array is empty is gone.

**Comment by rpherbig** [comment]
This PR has gotten a bit large (29 commits). We're starting to see some merge conflicts and fixing multiple things at once.

What would you think about issuing a separate PR for the loot changes? Then closing this PR and making a new PR for the necro changes (rebased on top of the new loot changes)?

Please don't take this the wrong way - we really appreciate your contribution :)

**Comment by jandersson** [comment]
None taken, this is a good learning experience for me. I'll need to read up a bit on how to rebase for a bit first, I don't have that in my git knowledge yet

**Comment by rcuhljr** [comment]
Record the sha of the commits you made on this branch you want to keep, move to a clean new branch, and `git cherry-pick shahash` to pull them over to that branch.
